src/Backend/ReasonML.idr

```diff
@@ -0,0 +1,82 @@
+module Backend.ReasonML
+
+import Data.Vect
+import Control.Monad.State
+
+import Backend.Utils
+import Types
+import Typedefs
+
+%default partial
+%access public export
+
+formatVar : String -> String
+formatVar = (strCons '\'') . lowercase
+
+-- TODO would be nice if type was `(e : Env n) -> Vect (fst (Vect.filter isLeft e)) String`, 
+-- since it then could be used by several backends. I'm having trouble getting it to typecheck.
+getFreeVars : Env n -> String
+getFreeVars e = withSep ", " (either formatVar (const "")) $ snd $ Vect.filter isLeft e
+
+makeType : Env n -> TDef n -> String
+makeType     _ T0            = "void"
+makeType     _ T1            = "unit"
+makeType {n} e (TSum xs)     = tsum xs
+  where
+  tsum : Vect (2 + _) (TDef n) -> String
+  tsum [x, y]              = "either" ++ (parens $ (makeType e x) ++ ", " ++ makeType e y)
+  tsum (x :: y :: z :: zs) = "either" ++ (parens $ (makeType e x) ++ ", " ++ tsum (y :: z :: zs))
+makeType     e (TProd xs)    = assert_total $ parens $ withSep ", " (makeType e) xs
+makeType     e (TVar v)      = either formatVar lowercase $ Vect.index v e
+makeType     e (TMu nam _)   = let freeVars = getFreeVars e
+                                in if freeVars == "" then lowercase nam else lowercase nam ++ parens freeVars
+makeType     _ (TName nam _) = lowercase nam
+
+makeDefs : Env n -> TDef n -> State (List Name) String
+makeDefs _ T0           = pure ""
+makeDefs _ T1           = pure ""
+makeDefs e (TProd xs)   = map concat $ traverse (makeDefs e) xs
+makeDefs e (TSum xs)    = map concat $ traverse (makeDefs e) xs
+makeDefs _ (TVar v)     = pure ""
+makeDefs e (TMu nam cs) = 
```

Comment:
Is the reason for abbreviating `name` to `nam` that t would cause shadowing?

Reply:
I think this is just mirroring what we did in the haskell one, and the reason there was basically following the academic style!

Reply:
I followed the conventions from the Haskell backend, renaming to `name` doesn't cause errors so I'll change it.